### PR TITLE
Add number of shards and replicas to `ElasticSearchConfig`

### PR DESCRIPTION
**Pull request by bugthunk**

### CODE REVIEW

**Comment by bugthunk** [comment]
Background: As the number of shards and replicas has to be controlled from the client due to Bloodhound, we (at Scrive) need to make this explicit in the configuration data type. Code that does not pattern match on `ElasticSearchConfig` will continue to work without modification, as will code that uses `defaultElasticSearchConfig`; the hard coded params of 4 shards and 1 replica is used.

Also, see https://github.com/bitemyapp/bloodhound/issues/227.

File: log-elasticsearch/src/Log/Backend/ElasticSearch/V1.hs [hunk 1]
```diff
@@ -96,11 +96,10 @@ elasticSearchLogger ElasticSearchConfig{..} genRandomWord = do
         -- index that already exists is harmless.
         indexExists' <- indexExists index
         unless indexExists' $ do
-          -- Bloodhound is weird and won't let us create index using default
-          -- settings, so pass these as the default ones.
+          -- Note that Bloodhound won't let us create index using default
```
Comment (23Skidoo) [inline log-elasticsearch/src/Log/Backend/ElasticSearch/V1.hs]: ...settings?

Comment (bugthunk) [reply log-elasticsearch/src/Log/Backend/ElasticSearch/V1.hs]: Eh - 'correctamundo'. Amending.

**Comment by jonathanjouty** [comment]
`bloodhound` really doesn't like ES defaults :D
* https://github.com/bitemyapp/bloodhound/pull/217
* https://github.com/bitemyapp/bloodhound/pull/120

**Comment by bugthunk** [comment]
@23Skidoo Can we merge this? I've another PR coming up. It's not exactly based on this but it would make things easier.

**Comment by 23Skidoo** [comment]
@bugthunk Sure, feel free to merge when (there's already a LGTM from me on this PR, btw). Do you need a new release?

**Comment by bugthunk** [comment]
Yes, I know about the 'LGTM'; I am just a bit chicken pressing the button myself :)
But I'll do it. And a new release would be great.

**Comment by bitemyapp** [comment]
@jonathanjouty I don't like losing data. Using auto-generated docids had caused data loss for me in the past. I'm debugging missing documents from bulk uploads to ES as we speak.

**Comment by jonathanjouty** [comment]
@bitemyapp understood.
I think the difference in perspective comes from the fact that _(1)_ we don't care that much about losing some data (ES is not our only log store) and _(2)_ we haven't been bitten by ES so much, largely due to lack of experience.
Anyway, thanks for maintaining Bloodhound, very appreciative despite our occasional gripes :)

**Comment by bitemyapp** [comment]
@jonathanjouty for context, I've been using ES in production off and on since ~2012. They didn't really have safe split brain merges until after Kyle got on them about it. I don't think that was fully fixed for a couple years after that though.

Some highlights:
- https://aphyr.com/posts/317-jepsen-elasticsearch
- https://aphyr.com/posts/323-call-me-maybe-elasticsearch-1-5-0

Further, you _do_ care about losing some data because if a critical log message is missing it could make debugging an error considerably harder. The current issue I'm figuring out in https://github.com/bitemyapp/katip (library that serves a similar purpose as `log`, I think. btw - what's the difference?) is an issue I've had off and on for years as have colleagues of mine: you bulk upload documents, only 30-70% of those documents show up in the index even after you force an index refresh. I'm currently uploading 10,000 documents 500 docs at a time and after I'm done and force a refresh I only see 3,500 to 6,500 documents. This is repeatable against an instance of Elasticsearch running on a machine where it is flush with resources. I remember fixing this before with the force refresh and putzing around with some index settings but nothing's making a difference this time around.

Over the course of a benchmark that inserts ~17mm documents, I'm only seeing 280-320k in the index at the end. That's not an occasional document lost.

**Comment by jonathanjouty** [comment]
@bitemyapp 
The data loss situation sounds worrying!
For now we've been OK, but we're new to using Elasticsearch.
Will keep eyes peeled for any issues...

> library that serves a similar purpose as `log`, I think. btw - what's the difference?

Haven't looked into details, but I know there is a difference in approach regarding the monadic context:

1. `katip` has `class MonadIO m => Katip m`.
2. `log` only has `class Monad m => MonadLog m`
  (_actually_ `MonadTime m`, which is `class Monad m => MonadTime m`)

   Then just an instance of `MonadIO`
   (again, _actually_ `(MonadBase IO m, MonadTime m) => MonadLog (LogT m)`)

First ~6 minutes of @arybczak's talk should explain it:
https://skillsmatter.com/skillscasts/8769-controlling-monadic-effects-with-type-classes

**Comment by bitemyapp** [comment]
I can't say that sifting out the ability to perform IO more granularly is going to help prevent problems in production when you're using Elasticsearch. Make sure you have a source of truth apart from ES that would permit replaying bulk inserts or doc updates against the ES index.

I'm not just speaking from my own experience here either, I have colleagues that have _huge_ ES deployments and it requires a tremendous amount of care to avoid data loss. It is nothing like using PostgreSQL or even MongoDB. Validate all of your inserts manually in the client!

**Pull request merged**